coordinator/src/position/models.rs

```diff
@@ -49,4 +56,198 @@ pub struct Position {
     /// This field is optional for backwards compatibility because we cannot deterministically
     /// associate already existing contracts with positions.
     pub temporary_contract_id: Option<ContractId>,
+    pub closing_price: Option<f32>,
+}
+
+impl Position {
+    /// Calculates the profit and loss for the coordinator
+    ///
+    /// The position stored represents the values of the trader.
+    pub fn calculate_coordinator_pnl(&self, quote: Quote) -> Result<i64> {
+        let closing_price = match self.closing_price {
+            None => closing_price(self.direction, quote),
+            Some(closing_price) => {
+                Decimal::try_from(closing_price).expect("f32 closing price to fit into decimal")
+            }
+        };
+
+        let average_entry_price = Decimal::try_from(self.average_entry_price)
+            .context("Failed to convert average entry price to Decimal")?;
+
+        let (long_leverage, short_leverage) = match self.direction {
+            Direction::Long => (self.leverage, COORDINATOR_LEVERAGE),
+            Direction::Short => (COORDINATOR_LEVERAGE, self.leverage),
+        };
+
+        // the position in the database is the trader's position, our direction is opposite
+        let direction = self.direction.opposite();
+
+        let pnl = calculate_pnl(
+            average_entry_price,
+            closing_price,
+            self.quantity,
+            long_leverage,
+            short_leverage,
+            direction,
+        )
+        .context("Failed to calculate pnl for position")?;
+
+        Ok(pnl)
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+    use std::str::FromStr;
+
+    #[test]
+    fn given_trader_long_position_when_no_bid_price_change_then_zero_pnl() {
+        let position = Position::dummy()
+            .with_leverage(2.0)
+            .with_quantity(1.0)
+            .with_average_entry_price(1000.0)
+            .with_direction(Direction::Long);
+
+        let quote = dummy_quote(1000, 0);
+
+        let coordinator_pnl = position.calculate_coordinator_pnl(quote).unwrap();
+
+        assert_eq!(coordinator_pnl, 0);
+    }
+
+    #[test]
+    fn given_trader_short_position_when_no_ask_price_change_then_zero_pnl() {
+        let position = Position::dummy()
+            .with_leverage(2.0)
+            .with_quantity(1.0)
+            .with_average_entry_price(1000.0)
+            .with_direction(Direction::Short);
+
+        let quote = dummy_quote(0, 1000);
+
+        let coordinator_pnl = position.calculate_coordinator_pnl(quote).unwrap();
+
+        assert_eq!(coordinator_pnl, 0);
+    }
+
+    /// See also: `given_long_position_when_price_10_pc_up_then_18pc_profit` test in `trade::cfd`
+    #[test]
+    fn given_trader_long_position_when_bid_price_10pc_up_then_coordinator_18pc_loss() {
```

Comment:
It would be good to also assert the pnl in percent given you mention it in the test function name.

Reply:
That is a good point, maybe *that* would help understanding it. It's still hard for me to grasp *why* it is 18.x percent though 😅

Reply:
It's definitely not intuitive, I agree. I assume you saw already this comment: https://github.com/get10101/10101/pull/967#discussion_r1272132643 which explains the 18%

Here is a simple calculator to play around with number if you haven't found it already: https://btctools.io/calculators/profit-loss

Reply:
I still don't get something. If the coordinator puts up 1 bitcoin (1 leverage) and they lose ~9M sats, isn't that a ~9% loss?

I think the 18 comes from considering the app side, which with leverage 2 doubles their ROE w.r.t leverage 1.

Reply:
That is correct, I was thinking from the taker's point of view. I'll change it to reflect the coordintor's pnl percentage.